gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/util/UserAgent.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.tinkerpop.gremlin.driver.util;
+
+import org.apache.tinkerpop.gremlin.util.Gremlin;
+import javax.naming.NamingException;
+
+public class UserAgent {
```

Comment:
Thanks for all of your feedback Divij, could I ask for some more elaboration of your expectations regarding decoupling the driver and server code for the user agent? Keeping things decoupled was something I kept in mind while implementing my changes and I'm not sure what further changes you are looking for in that regard. As it stands all the code which generates and sends the user agent exists in gremlin-driver (which will be re-implemented in each GLV in a future PR). All the code which extracts the user agent, logs it for debugging, and adds it to the server metrics is wrapped into `WsUserAgentHandler` in gremlin-server. The only use of driver code from the server currently (excluding tests) is that the server uses the driver's constant for the "User-Agent" header name. This constant could easily be duplicated in the server to avoid referencing driver code from the server.

Reply:
I was thinking in the following direction:

1. UserAgent.java is a model (POJO) that describes the UserAgent. It contains fields such as JavaVersion etc. along with the getter & setters. It contains methods "serializeToString" and "deserializeFromString". This model resides in a place which is common to all drivers, communication protocols and server. I would suggest placing it in https://github.com/apache/tinkerpop/tree/master/gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver at same level as Host.java. This POJO/modelling for UserAgent would help us from code extensibility perspective in future if we want to add enums or constraints checks on values of user agent.

2. Both HTTP and WebSocket will create this POJO using a static method in the above file, perhaps, UserAgent#Generate() 


The goal of this structure is to ensure that code is extensible to HTTP in future and that individual methods could be unit tested easily. I don't have a strong opinion on this except for things such as WS_HANDSHAKE_USER_AGENT which should be made agnostic of WebSocket (the same could be used for HTTP).

Reply:
Thanks for responding so soon. I agree that all of these changes would provide future extensibility benefits and I especially agree with the naming change to be web socket agnostic. I have pushed a small change where I have moved the UserAgent class to your suggested location and renamed WS_HANDSHAKE_USER_AGENT to just USER_AGENT. I have opted to keep USER_AGENT as a public statically initialized string constant as it is the most lightweight method of accessing the user agent and it meets the current needs. I certainly would not oppose modifying this further in the manner you've suggested sometime in the future if additional extensibility is required.